### ISSUE TITLE
nextJS API route.ts stuck on await request.formData()

### ISSUE THREAD

**Issue by ishaangandhi** [vercel/next.js#64002]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/condescending-haze-6f3k9w

### To Reproduce

Upload a multipart form data and wait on it. This works locally, but when you deploy it to Vercel, it just hangs. If you switch to JSON instead of multipart form data, it works fine. So its not an issue of the bytes taking too long in flight.

### Current vs. Expected behavior

I expected the `await request.formData()` to take as long as a similarly sized `request.json()`, but it regularly times out.

### Provide environment information

```bash
Downloading registry.npmjs.org/next/14.1.4: 20.77 MB/20.77 MB, done
Downloading registry.npmjs.org/@next/swc-darwin-x64/14.1.4: 35.35 MB/35.35 MB, done
Progress: resolved 28, reused 16, downloaded 4, added 20, done

Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 23.4.0: Wed Feb 21 21:45:49 PST 2024; root:xnu-10063.101.15~2/RELEASE_ARM64_T6020
Binaries:
  Node: 21.6.2
  npm: 10.2.4
  Yarn: N/A
  pnpm: 8.12.1
Relevant Packages:
  next: 14.1.4
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: N/A
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router

### Which stage(s) are affected? (Select all that apply)

Vercel (Deployed)

**Comment by ishaangandhi** [comment]
Found someone else with the same issue on SO: https://stackoverflow.com/questions/78149924/nextjs-api-route-ts-stuck-on-await-request-formdata

**Comment by mertthesamael** [comment]
Well, it's not a solid "solution" for it and I don't know that if it fits in your project but maybe you may want to try [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#forms) to interact with FormData

**Comment by ishaangandhi** [comment]
That's pretty cool @mertthesamael. I'll definitely use that the next time I make a form.

**Comment by antoinedelplace** [comment]
I have the same problem. Any solution yet ?

**Comment by ishaangandhi** [comment]
> I have the same problem. Any solution yet ?
> 
> _Edit by maintainer bot: Comment was **automatically** minimized because it was considered unhelpful. (If you think this was by mistake, let us know). Please only comment if it adds context to the issue. If you want to express that you have the same problem, use the upvote 👍 on the issue description or subscribe to the issue for updates. Thanks!_

I'd recommend changing to Server actions.

**Comment by PaulSinghDev** [comment]
Just adding some context here as I have experienced the same issue. These are my findings so far:

1. The function works fine in the Vercel prod env
2. The function works fine when run locally
3. The function works fine when I use a vite test to send the request (locally)
4. The function falls down when I try to run the aforementioned test in a GitHub Action Workflow
5. The function falls down when I try to locally run the aforementioned test in a GitHub Action Workflow via act

The workflow is using `ubuntu-latest` as the machine it is run on so I am starting to think  there is a nuance with the operating system which is causing the issue. 

I checked the linked SO question but it seems to lean toward the issue being the file's size. I do not believe this to be the case as I am passing a file which is tiny (16bytes) and am generating it like this:

```
new File(
      [new Blob(["0x00", "0x01", "0x02", "0x03"], { type: "image/png" })],
      "test.png",
      { type: "image/png" }
    )
    ```
    
The weird thing is, when I run the test locally (not within the GH action) the files size pre request is `16 bytes` but, when it hits the server, logging `file.size` returns `0`. 

I understand the notion to use server actions but that would require ripping out a lot of what has been written already for a large CRUD app so would be nice to have this resolved.

**Comment by Connoropolous** [comment]
try investigating your nodejs version. switching from v18 to v20.18 just fixed the same ‘hanging forever’ issue for me. 
see here:
https://github.com/vercel/next.js/issues/52616#issuecomment-1749824093

**Comment by raffidahmad** [comment]
still an issue in node v22 with expo api routes

**Comment by marcoshernanz** [comment]
Thanks for the report. I reproduced this with Next.js 14.1.4 on Node 18, but not on Node 20, the reported local Node 21.6.2, or current Node 24. The same multipart body also stalls in Node 18's native `Request.formData()` without Next.js, which points to the old Node runtime rather than Next.js.

I verified the case on `next@canary` locally and on Vercel, where it completes normally. Node 18 is no longer supported by current Next.js or Vercel, and the linked reproduction no longer contains the failing route, so I am closing this as resolved by the supported runtime stack. If this still happens, please open a new issue with a minimal `next@canary` reproduction and the deployed Node version.
